### ISSUE TITLE
No events are fired on windows

### ISSUE THREAD

**Issue by oumad** [Axosoft/nsfw#131]
We've been using chokidar for our app so far, but we keep getting performance/stability issues. We recently discovered this library and we wanted to give it a go. But so far I couldn't get any event fired. Am I using this correctly ?

```js
var nsfw = require("nsfw");

var watcher1;
return nsfw(
  "C:/dev/testing/input",
  function (events) {
    // handle events
    console.log("events", events);
  },
  {
    debounceMS: 250,
    errorCallback(errors) {
      console.log("callback error", errors);
    },
  }
)
  .then(function (watcher) {
    watcher1 = watcher;
    // console.log("starting");
    return watcher1.start();
  })
  .then(function () {
    // we are now watching dir1 for events!
    console.log("watching now...");
    // To stop watching
    //watcher1.stop();
  })
  .catch((err) => {
    console.log("catched error", err);
  });
```

The only output I get so far is my console.log 
`watching now...`

Also sometimes randomly when I run the code I get my promise catch error
`catched error [Error: NSFW was unable to start watching that directory.]`
Then I restart and it goes back to the `watching now...`, I try to delete, rename, create, move files in the directory that I specified, but no event is fired.

I am using windows 10
node : 12.21.0
"nsfw": "^2.1.1"

**Comment by implausible** [comment]
Can you tell me a little more about your system? I am unable to reproduce your issue on Windows 10 and we definitely have specs that confirm that this module can watch files in a directory. What version of windows are you on? Version of Node? Version of NSFW?

**Comment by oumad** [comment]
> If I'm reading your code correctly, you start the watcher and then _immediately shut the watcher down_ using `watcher1.stop()`. Of course it doesn't report any file changes 😄

Oh actually, that's a part of the code that I put in there when I thought there must be sth wrong in the way I'm using the lib, so I basically copied exactly the example code in the README, except I got the same behavior even before that `watcher1.stop()` line. I just tested again, commented out that part, still getting same behavior ^^'.

**Comment by implausible** [comment]
What version of NSFW? What version of Node? Version of Windows? Are there any special qualities about the folder you're watching (`C:/dev/testing/input` is a folder, not a symlink or file)?

Can you write a test case demonstrating the problem and submit it as a PR?

**Comment by oumad** [comment]
Ok with more testing we found that it's a problem on my own machine. I still couldn't figure out why, I tested different versions of node. For now the issue can be closed since it is not directly related to nsfw.

**Comment by marechal-p** [comment]
@oumad It's not just you, I am having troubles on Windows too. Can you run the test suite from this repository?

I am running Windows 10 version 20H2 build 19042.867.

<details><summary>

_`yarn test` output from [35aa0eaf1104202726387e316d1f0315b7ff7143](https://github.com/Axosoft/nsfw/tree/35aa0eaf1104202726387e316d1f0315b7ff7143)_

</summary>

```
  Node Sentinel File Watcher
    Basic
      √ can watch a single file (12038ms)
      1) can listen for a create event
      2) can listen for a delete event
      3) can listen for a modify event
      4) can listen for a rename event
      5) can listen for a move event
      6) can run multiple watchers at once
      √ will properly track the movement of watched directories across watched directories (6080ms)
    Recursive
      7) can listen for the creation of a deeply nested file
      8) can listen for the destruction of a directory and its subtree
      9) does not loop endlessly when watching directories with recursive symlinks
    Pausing
      10) can pause and resume event reporting
    Errors
      11) can gracefully recover when the watch folder is deleted
    Stress
Cloning into 'nsfw-stress-test'...
Updating files: 100% (5833/5833), done.
      12) does not segfault under stress
Cloning into 'nsfw-stress-test'...
Updating files: 100% (5833/5833), done.
      13) creates and destroys many watchers
    Unicode support
      14) supports watching unicode directories
    Garbage collection
      15) can garbage collect all instances


  2 passing (4m)
  15 failing

  1) Node Sentinel File Watcher
       Basic
         can listen for a create event:
     Error: This NSFW cannot be stopped, because it is not running.


  2) Node Sentinel File Watcher
       Basic
         can listen for a delete event:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(eventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:162:16)

  3) Node Sentinel File Watcher
       Basic
         can listen for a modify event:
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (D:\sources\nsfw\js\spec\index-spec.js)
      at listOnTimeout (internal/timers.js:549:17)
      at processTimers (internal/timers.js:492:7)

  4) Node Sentinel File Watcher
       Basic
         can listen for a rename event:
     Error: This NSFW cannot be stopped, because it is not running.


  5) Node Sentinel File Watcher
       Basic
         can listen for a move event:
     Error: This NSFW cannot be stopped, because it is not running.


  6) Node Sentinel File Watcher
       Basic
         can run multiple watchers at once:

      AssertionError [ERR_ASSERTION]: 0 == 2
      + expected - actual

      -0
      +2

      at Context.<anonymous> (js\spec\index-spec.js:397:16)

  7) Node Sentinel File Watcher
       Recursive
         can listen for the creation of a deeply nested file:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(foundFileCreateEvent)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:475:16)

  8) Node Sentinel File Watcher
       Recursive
         can listen for the destruction of a directory and its subtree:
     Error: This NSFW cannot be stopped, because it is not running.


  9) Node Sentinel File Watcher
       Recursive
         does not loop endlessly when watching directories with recursive symlinks:
     Error: EPERM: operation not permitted, symlink 'D:\sources\nsfw\mockfs\test' -> 'D:\sources\nsfw\mockfs\test\link'


  10) Node Sentinel File Watcher
       Pausing
         can pause and resume event reporting:
     Error: This NSFW cannot be stopped, because it is not running.


  11) Node Sentinel File Watcher
       Errors
         can gracefully recover when the watch folder is deleted:
     Error: This NSFW cannot be stopped, because it is not running.


  12) Node Sentinel File Watcher
       Stress
         does not segfault under stress:
     Error: This NSFW cannot be stopped, because it is not running.


  13) Node Sentinel File Watcher
       Stress
         creates and destroys many watchers:
     Error: NSFW was unable to start watching that directory.


  14) Node Sentinel File Watcher
       Unicode support
         supports watching unicode directories:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(eventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:721:16)

  15) Node Sentinel File Watcher
       Garbage collection
         can garbage collect all instances:
     Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (D:\sources\nsfw\js\spec\index-spec.js)
      at listOnTimeout (internal/timers.js:549:17)
      at processTimers (internal/timers.js:492:7)
```

</details>

**Comment by implausible** [comment]
@marechal-p any chance you can bubble up the error from starting the watcher on windows to see what windows is not happy about?

**Comment by oumad** [comment]
@marechal-p What worked for me is that I used another machine. So far I still can't get it to work on my own machine.

I just run the repo test on my machine and this is what I got : 

<details><summary>Test output</summary>
<p>

```
Node Sentinel File Watcher
    Basic
      √ can watch a single file (12032ms)
      1) can listen for a create event
      2) can listen for a delete event
      3) can listen for a modify event
      4) can listen for a rename event
      5) can listen for a move event
      6) can run multiple watchers at once
      √ will properly track the movement of watched directories across watched directories (6074ms)
    Recursive
      7) can listen for the creation of a deeply nested file
      8) can listen for the destruction of a directory and its subtree
      √ does not loop endlessly when watching directories with recursive symlinks
    Pausing
      9) can pause and resume event reporting
    Errors
      10) can gracefully recover when the watch folder is deleted
    Stress
Cloning into 'nsfw-stress-test'...
Updating files: 100% (5833/5833), done.
      11) does not segfault under stress
Cloning into 'nsfw-stress-test'...
Updating files: 100% (5833/5833), done.
      12) creates and destroys many watchers
    Unicode support
      13) supports watching unicode directories
    Garbage collection
      14) can garbage collect all instances


  3 passing (7m)
  14 failing

  1) Node Sentinel File Watcher
       Basic
         can listen for a create event:
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (C:\dev\testijs\spec\index-spec.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)

  2) Node Sentinel File Watcher
       Basic
         can listen for a delete event:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(eventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:162:16)

  3) Node Sentinel File Watcher
       Basic
         can listen for a modify event:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert(eventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:196:9)

  4) Node Sentinel File Watcher
       Basic
         can listen for a rename event:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(renameEventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:268:22)

  5) Node Sentinel File Watcher
       Basic
         can listen for a move event:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(deleteEventFound && createEventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:346:20)

  6) Node Sentinel File Watcher
       Basic
         can run multiple watchers at once:

      AssertionError [ERR_ASSERTION]: 0 == 2
      + expected - actual

      -0
      +2

      at Context.<anonymous> (js\spec\index-spec.js:397:16)

  7) Node Sentinel File Watcher
       Recursive
         can listen for the creation of a deeply nested file:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(foundFileCreateEvent)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:475:16)

  8) Node Sentinel File Watcher
       Recursive
         can listen for the destruction of a directory and its subtree:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(deletionCount > 2)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:513:16)

  9) Node Sentinel File Watcher
       Pausing
         can pause and resume event reporting:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(eventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:580:16)

  10) Node Sentinel File Watcher
       Errors
         can gracefully recover when the watch folder is deleted:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(erroredOut)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:604:16)

  11) Node Sentinel File Watcher
       Stress
         does not segfault under stress:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(count > 0)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:639:16)

  12) Node Sentinel File Watcher
       Stress
         creates and destroys many watchers:
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (C:\dev\testing\nsfw\js\spec\index-spec.js)        
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)

  13) Node Sentinel File Watcher
       Unicode support
         supports watching unicode directories:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert.ok(eventFound)

      + expected - actual

      -false
      +true

      at Context.<anonymous> (js\spec\index-spec.js:721:16)

  14) Node Sentinel File Watcher
       Garbage collection
         can garbage collect all instances:
     Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (C:\dev\testing\nsfw\js\spec\index-spec.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)

```

  

</p>
</details>

I am running Microsoft Windows 10 Pro Version10.0.19042 Build 19042 node v12.21.0

**Comment by marechal-p** [comment]
@oumad is the machine that doesn't work running on a Ryzen processor, and the one that works on Intel?

**Comment by marechal-p** [comment]
@oumad also please tell me if the following patch fixes the issue for you too: https://github.com/Axosoft/nsfw/pull/133

**Comment by oumad** [comment]
> @oumad also please tell me if the following patch fixes the issue for you too: #133

I just tested the `theia-ide:mp/win32-race` branch, and it works on my machine ! I re-tested with the master release, to make sure. 
Definitely #133 fixes the issue for me ! I hope it'll get merged soon.

As for the CPU, my machine is on Intel Xeon E5-2660. The other machines where it worked are on Intel too.

**Comment by implausible** [comment]
v2.1.2 has been published. Excellent work @marechal-p.
